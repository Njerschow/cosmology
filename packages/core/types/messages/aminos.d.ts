export namespace aminos {
    namespace createPool {
        function toAmino(): void;
        function fromAmino(): void;
    }
    namespace joinPool {
        export function toAmino_1({ sender, poolId, shareOutAmount, tokenInMaxs }: {
            sender: any;
            poolId: any;
            shareOutAmount: any;
            tokenInMaxs: any;
        }): {
            sender: any;
            poolId: any;
            shareOutAmount: any;
            tokenInMaxs: any;
        };
        export { toAmino_1 as toAmino };
        export function fromAmino_1({ sender, poolId, shareOutAmount, tokenInMaxs }: {
            sender: any;
            poolId: any;
            shareOutAmount: any;
            tokenInMaxs: any;
        }): {
            sender: any;
            poolId: any;
            shareOutAmount: any;
            tokenInMaxs: any;
        };
        export { fromAmino_1 as fromAmino };
    }
    namespace joinSwapExternAmountIn {
        export function toAmino_2(): void;
        export { toAmino_2 as toAmino };
        export function fromAmino_2(): void;
        export { fromAmino_2 as fromAmino };
    }
    namespace exitPool {
        export function toAmino_3(): void;
        export { toAmino_3 as toAmino };
        export function fromAmino_3(): void;
        export { fromAmino_3 as fromAmino };
    }
    namespace swapExactAmountIn {
        export function toAmino_4({ sender, routes, tokenIn, tokenOutMinAmount }: {
            sender: any;
            routes: any;
            tokenIn: any;
            tokenOutMinAmount: any;
        }): {
            sender: any;
            routes: any;
            tokenIn: {
                denom: any;
                amount: any;
            };
            tokenOutMinAmount: any;
        };
        export { toAmino_4 as toAmino };
        export function fromAmino_4({ sender, routes, tokenIn, tokenOutMinAmount }: {
            sender: any;
            routes: any;
            tokenIn: any;
            tokenOutMinAmount: any;
        }): {
            sender: any;
            routes: any;
            tokenIn: any;
            tokenOutMinAmount: any;
        };
        export { fromAmino_4 as fromAmino };
    }
    namespace swapExactAmountOut {
        export function toAmino_5(): void;
        export { toAmino_5 as toAmino };
        export function fromAmino_5(): void;
        export { fromAmino_5 as fromAmino };
    }
    namespace lockTokens {
        export function toAmino_6({ owner, duration, coins }: {
            owner: any;
            duration: any;
            coins: any;
        }): {
            owner: any;
            coins: any;
            duration: string;
        };
        export { toAmino_6 as toAmino };
        export function fromAmino_6({ owner, duration, coins }: {
            owner: any;
            duration: any;
            coins: any;
        }): {
            owner: any;
            coins: any;
            duration: {
                seconds: any;
                nanos: number;
            };
        };
        export { fromAmino_6 as fromAmino };
    }
    namespace beginUnlocking {
        export function toAmino_7(): void;
        export { toAmino_7 as toAmino };
        export function fromAmino_7(): void;
        export { fromAmino_7 as fromAmino };
    }
    namespace unlockPeriodLock {
        export function toAmino_8(): void;
        export { toAmino_8 as toAmino };
        export function fromAmino_8(): void;
        export { fromAmino_8 as fromAmino };
    }
}
