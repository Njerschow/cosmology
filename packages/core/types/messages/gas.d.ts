export namespace gas {
    namespace MsgSend {
        namespace uosmo {
            const amount: number;
            const gas: string;
        }
        namespace uatom {
            const amount_1: number;
            export { amount_1 as amount };
            const gas_1: string;
            export { gas_1 as gas };
        }
    }
    namespace MsgTransfer {
        export namespace uosmo_1 {
            const amount_2: number;
            export { amount_2 as amount };
            const gas_2: string;
            export { gas_2 as gas };
        }
        export { uosmo_1 as uosmo };
        export namespace uatom_1 {
            const amount_3: number;
            export { amount_3 as amount };
            const gas_3: string;
            export { gas_3 as gas };
        }
        export { uatom_1 as uatom };
    }
}
