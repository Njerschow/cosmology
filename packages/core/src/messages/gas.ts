export const gas = {
  osmosis: {
    createPool: {
      gas: 250000
    },
    joinPool: {
      gas: 140000,
      shareCoinDecimals: 18
    },
    joinSwapExternAmountIn: {
      gas: 140000,
      shareCoinDecimals: 18
    },
    exitPool: {
      gas: 140000,
      shareCoinDecimals: 18
    },
    swapExactAmountIn: {
      gas: 250000
    },
    swapExactAmountOut: {
      gas: 250000
    },
    lockTokens: {
      gas: 250000
    },
    beginUnlocking: {
      gas: 140000
    },
    unlockPeriodLock: {
      gas: 140000
    }
  }
};
