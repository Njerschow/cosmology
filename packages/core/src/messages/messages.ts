import { osmosis } from '@osmonauts/osmosis/main/proto/osmosis';

export const messages = {
  ...osmosis.gamm.v1beta1.json,
  ...osmosis.superfluid.json,
  ...osmosis.lockup.json
};
