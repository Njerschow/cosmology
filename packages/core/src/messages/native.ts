import { cosmos, ibc } from '@cosmonauts/osmosis';

export const native = {
  ...cosmos.staking.v1beta1.messages,
  ...cosmos.distribution.v1beta1.messages,
  ...cosmos.gov.v1beta1.messages,
  ...cosmos.bank.v1beta1.messages,
  ...ibc.applications.transfer.v1.messages
};
