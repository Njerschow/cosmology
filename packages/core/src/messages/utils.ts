import {
  AminoTypes,
  calculateFee,
  GasPrice,
  SigningStargateClient
} from '@cosmjs/stargate';
import { Registry } from '@cosmjs/proto-signing';
import { TxRaw } from 'cosmjs-types/cosmos/tx/v1beta1/tx';
import { coins } from '@cosmjs/amino';
import { defaultRegistryTypes } from '@cosmjs/stargate';
import retry from 'retry';
import { aminos } from './aminos';
import { meta as metaInfo } from './meta';
import { Dec, IntPretty } from '@keplr-wallet/unit';
import { BroadcastTxResponse } from '../types';
import { OfflineSigner } from '@cosmjs/proto-signing'

// import { osmosis } from '@cosmonauts/osmosis';

export const getSigningOsmosisClient = async ({ rpcEndpoint, signer }: { rpcEndpoint: string, signer: OfflineSigner }) => {
  // registry
  const registry = new Registry(defaultRegistryTypes);

  // const additions = {
  //   ...osmosis.gamm.v1beta1.AminoConverter,
  //   ...osmosis.lockup.AminoConverter,
  //   ...osmosis.superfluid.AminoConverter
  // };

  // aminotypes
  const aminoTypes = new AminoTypes({
    additions: Object.keys(aminos).reduce((m, key) => {
      const meta = metaInfo[key];
      const { toAmino, fromAmino } = aminos[key];
      m[meta.amino] = {
        aminoType: meta.type,
        toAmino,
        fromAmino
      };
      return m;
    }, {})
  });

  // register the goods
  Object.keys(aminos).forEach((key) => {
    const meta = metaInfo[key];
    registry.register(meta.amino, meta.osmosis);
  });

  const client = await SigningStargateClient.connectWithSigner(
    rpcEndpoint,
    signer,
    { registry: registry, aminoTypes: aminoTypes }
  );

  return client;
};

export const signAndBroadcast = async ({
  client,
  chainId,
  address,
  msg,
  fee,
  memo = ''
}: {
  client: SigningStargateClient,
  chainId: string,
  address: string,
  msg: any,
  fee: any,
  memo: string
}): Promise<BroadcastTxResponse> => {
  const { accountNumber, sequence } = await client.getSequence(address);
  const txRaw = await client.sign(address, [msg], fee, memo, {
    accountNumber: accountNumber,
    sequence: sequence,
    chainId
  });
  const txBytes = TxRaw.encode(txRaw).finish();
  return await client.broadcastTx(txBytes);
};

export const signAndBroadcastBatch = async ({
  client,
  chainId,
  address,
  msgs,
  fee,
  memo = ''
}: {
  client: SigningStargateClient,
  chainId: string,
  address: string,
  msgs: any[],
  fee: any,
  memo: string
}) => {
  const { accountNumber, sequence } = await client.getSequence(address);
  const txRaw = await client.sign(address, msgs, fee, memo, {
    accountNumber: accountNumber,
    sequence: sequence,
    chainId
  });
  const txBytes = TxRaw.encode(txRaw).finish();
  return await client.broadcastTx(txBytes);
};

export const getOsmoFee = (name) => {
  if (!metaInfo[name]) throw new Error('missing message.');
  const gas = metaInfo[name].gas + '';
  const fee = {
    amount: coins(0, 'uosmo'),
    gas
  };
  return fee;
};

export const estimateOsmoFee = async (client: SigningStargateClient, address: string, msgs: any[], memo: string) => {
  const gasPrice = GasPrice.fromString('0.025uosmo');
  const gasEstimation = await client.simulate(address, msgs, memo);
  const fee = calculateFee(Math.round(gasEstimation * 1.3), gasPrice);
  return fee;
};

export const noDecimals = (num) => {
  return new IntPretty(new Dec(num)).maxDecimals(0).locale(false).toString();
};
