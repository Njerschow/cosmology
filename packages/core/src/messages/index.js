export * from './meta';
export * from './aminos';
export * from './utils';

import { messages as msgs } from './messages';
import { native } from './native';

export const messages = {
  ...msgs,
  ...native
};
