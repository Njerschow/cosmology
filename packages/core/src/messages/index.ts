export * from './gas';
export * from './utils';

import { messages as msgs } from './messages';
import { native } from './native';

export const messages = {
  ...msgs,
  ...native
};

import * as telescope from '@osmonauts/osmosis';

export {
  telescope
}