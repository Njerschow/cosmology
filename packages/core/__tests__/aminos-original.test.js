// @ts-nocheck
import { aminos } from '../src/messages/aminos';
import { meta as metaInfo } from '../src/messages/meta';

it('aminos', async () => {
  const additions = Object.keys(aminos).reduce((m, key) => {
    const meta = metaInfo[key];
    const { toAmino, fromAmino } = aminos[key];
    m[meta.amino] = {
      aminoType: meta.type,
      toAmino,
      fromAmino
    };
    return m;
  }, {});
  expect(additions).toMatchSnapshot();
});
it('lockTokens.fromAmino', async () => {
  const fromAmino = aminos.lockTokens.fromAmino({
    owner: 'osmo1addresshere',
    coins: [
      {
        denom: 'mydenom',
        amount: '1000'
      }
    ],
    duration: {
      seconds: '100',
      nanos: 0
    }
  });
  expect(fromAmino).toMatchSnapshot();
});
it('lockTokens.toAmino', async () => {
  const toAmino = aminos.lockTokens.toAmino({
    owner: 'osmo1addresshere',
    coins: [
      {
        denom: 'mydenom',
        amount: '1000'
      }
    ],
    duration: {
      seconds: '100',
      nanos: 0
    }
  });
  expect(toAmino).toMatchSnapshot();
});

it('lockTokens multiple aminos', async () => {
  const toAmino = aminos.lockTokens.toAmino({
    owner: 'osmo1addresshere',
    coins: [
      {
        denom: 'mydenom',
        amount: '1000'
      }
    ],
    duration: {
      seconds: '100',
      nanos: 0
    }
  });
  expect(toAmino).toMatchSnapshot();
  const fromAmino = aminos.lockTokens.fromAmino(toAmino);
  expect(fromAmino).toMatchSnapshot();
});
