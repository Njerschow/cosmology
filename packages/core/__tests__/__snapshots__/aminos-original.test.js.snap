// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aminos 1`] = `
Object {
  "/osmosis.gamm.v1beta1.MsgCreatePool": Object {
    "aminoType": "osmosis/gamm/create-pool",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.gamm.v1beta1.MsgExitPool": Object {
    "aminoType": "osmosis/gamm/exit-pool",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.gamm.v1beta1.MsgJoinPool": Object {
    "aminoType": "osmosis/gamm/join-pool",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.gamm.v1beta1.MsgJoinSwapExternAmountIn": Object {
    "aminoType": "osmosis/gamm/join-swap-extern-amount-in",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.gamm.v1beta1.MsgSwapExactAmountIn": Object {
    "aminoType": "osmosis/gamm/swap-exact-amount-in",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.gamm.v1beta1.MsgSwapExactAmountOut": Object {
    "aminoType": "osmosis/gamm/swap-exact-amount-out",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.lockup.MsgBeginUnlocking": Object {
    "aminoType": "osmosis/lockup/begin-unlock-period-lock",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "/osmosis.lockup.MsgLockTokens": Object {
    "aminoType": "osmosis/lockup/lock-tokens",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
  "undefined": Object {
    "aminoType": "osmosis/lockup/unlock-period-lock",
    "fromAmino": [Function],
    "toAmino": [Function],
  },
}
`;

exports[`lockTokens multiple aminos 1`] = `
Object {
  "coins": Array [
    Object {
      "amount": "1000",
      "denom": "mydenom",
    },
  ],
  "duration": "NaN",
  "owner": "osmo1addresshere",
}
`;

exports[`lockTokens multiple aminos 2`] = `
Object {
  "coins": Array [
    Object {
      "amount": "1000",
      "denom": "mydenom",
    },
  ],
  "duration": Object {
    "nanos": NaN,
    "seconds": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
  },
  "owner": "osmo1addresshere",
}
`;

exports[`lockTokens.fromAmino 1`] = `
Object {
  "coins": Array [
    Object {
      "amount": "1000",
      "denom": "mydenom",
    },
  ],
  "duration": Object {
    "nanos": NaN,
    "seconds": Long {
      "high": 0,
      "low": 0,
      "unsigned": false,
    },
  },
  "owner": "osmo1addresshere",
}
`;

exports[`lockTokens.toAmino 1`] = `
Object {
  "coins": Array [
    Object {
      "amount": "1000",
      "denom": "mydenom",
    },
  ],
  "duration": "NaN",
  "owner": "osmo1addresshere",
}
`;
